google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/CookiesHolderTest.java

```diff
@@ -0,0 +1,551 @@
+                  @Override
+                  public void sendHeaders(Metadata headers) {
+                    headers.put(routingCookieKey, routingCookieValue);
+                    headers.put(ROUTING_COOKIE_2, "will-be-overridden");
+                    super.sendHeaders(headers);
+                  }
+                },
+                metadata);
+          }
+        };
+
+    Server newServer = FakeServiceBuilder.create(fakeService).intercept(serverInterceptor).start();
+
+    BigtableDataSettings.Builder settings =
+        BigtableDataSettings.newBuilderForEmulator(newServer.getPort())
+            .setProjectId("fake-project")
+            .setInstanceId("fake-instance");
+
+    BigtableDataClient client = BigtableDataClient.create(settings.build());
+
+    client.readRows(Query.create("table")).iterator().hasNext();
+
+    Metadata lastMetadata = serverMetadata.get(fakeService.count.get() - 1);
+
+    assertThat(lastMetadata)
+        .containsAtLeast(
+            ROUTING_COOKIE_2.name(), testCookie, routingCookieKey.name(), routingCookieValue);
+
+    newServer.shutdown();
+  }
+
+  @Test
+  public void testAllMethodsAreCalled() throws InterruptedException {
+    // This test ensures that all methods respect the retry cookie except for the ones that are
+    // explicitly added to the methods list. It requires that any newly method is exercised in this
+    // test. This is enforced by introspecting grpc method descriptors.
+    client.readRows(Query.create("fake-table")).iterator().hasNext();
+
+    fakeService.count.set(0);
+    client.mutateRow(RowMutation.create("fake-table", "key").setCell("cf", "q", "v"));
+
+    fakeService.count.set(0);
+    client.bulkMutateRows(
+        BulkMutation.create("fake-table")
+            .add(RowMutationEntry.create("key").setCell("cf", "q", "v")));
+
+    fakeService.count.set(0);
+    client.sampleRowKeys("fake-table");
+
+    fakeService.count.set(0);
+    client.checkAndMutateRow(
+        ConditionalRowMutation.create("fake-table", "key")
+            .then(Mutation.create().setCell("cf", "q", "v")));
+
+    fakeService.count.set(0);
+    client.readModifyWriteRow(
+        ReadModifyWriteRow.create("fake-table", "key").append("cf", "q", "v"));
+
+    fakeService.count.set(0);
+    client.generateInitialChangeStreamPartitions("fake-table").iterator().hasNext();
+
+    fakeService.count.set(0);
+    client.readChangeStream(ReadChangeStreamQuery.create("fake-table")).iterator().hasNext();
+
+    Set<String> expected =
+        BigtableGrpc.getServiceDescriptor().getMethods().stream()
+            .map(MethodDescriptor::getBareMethodName)
+            .collect(Collectors.toSet());
+
+    // Exclude methods that are not supported by routing cookie
+    methods.add("PingAndWarm");
+
+    assertThat(methods).containsExactlyElementsIn(expected);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    client.close();
+    server.shutdown();
```

Comment:
just to avoid debugging headaches later, can you wrap these in if statement that checks that the resource is not null.
This will help debug problems if an exception is thrown in setup()